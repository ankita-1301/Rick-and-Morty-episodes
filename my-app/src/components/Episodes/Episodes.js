import React from "react";
import "../../styles.css";
import "antd/dist/antd.css";
import { Card, Button } from "antd";
import { Link } from "react-router-dom";
import { getEpisode } from "rickmortyapi";

//Component for rendering all the episodes
export default class Episodes extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      episodes: [],
      next: 0, //counter if nextPage exists for all episodes
    };
  }
  componentDidMount() {
    this.fetchAllEpisodes();
  }

  //get all the episodes and update its next page with a counter
  fetchAllEpisodes = () => {
    getEpisode()
      .then(({ info, results }) => {
        this.setState({
          episodes: results,
          next: !!info.next ? this.state.next + 1 : this.state.next,
        });
      })
      .catch((err) => console.log(err));
  };

  //load more episodes when click on button and update next counter
  onClickLoadMore = () => {
    getEpisode({ page: this.state.next + 1 })
      .then(({ info, results }) => {
        this.setState({ episodes: [...this.state.episodes, ...results] });
        this.setState({ next: !!info.next ? this.state.next + 1 : 0 });
      })
      .catch((err) => console.log(err));
  };

  render() {
    const { episodes, next } = this.state;
    const gridStyle = {
      width: "25%",
      textAlign: "center",
      height: "120px",
      padding: "10px",
    };

    //card-grid for all the episodes with its link to EpisodeView
    let cardData = episodes.map((data, i) => {
      return (
        <Link
          to={{
            pathname: `/episodes/episode/${data.id}`,
            state: { data: data }, //episode details
          }}
          key={data.id}
        >
          <Card.Grid style={gridStyle} key={data.id}>
            <h1>{data.episode}</h1>
            <h3>{data.name}</h3>
          </Card.Grid>
        </Link>
      );
    });

    return (
      <div>
        <Card title="All episodes of Ricky and Morty" data-test="episodes-card">
          {cardData}
          {!!next && (
            <Button className="load-more-button" onClick={this.onClickLoadMore}>
              Load more...
            </Button>
          )}
        </Card>
      </div>
    );
  }
}
