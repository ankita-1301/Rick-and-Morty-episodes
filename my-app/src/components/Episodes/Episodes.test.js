import React from "react";
import Enzyme, { shallow } from "enzyme";
import Episodes from "./Episodes";
import renderer from "react-test-renderer";
import Adapter from "enzyme-adapter-react-16";

Enzyme.configure({ adapter: new Adapter() });

//test setup for episodes component

test("Episodes component snapshot", () => {
  const characters = renderer.create(<Episodes />);
  let charTree = characters.toJSON();
  expect(charTree).toMatchSnapshot();
});

const findByAttri = (component, attri) => {
  const wrapper = component.find(`[data-test="${attri}"]`);
  return wrapper;
};

const setUp = (props = {}) => {
  const component = shallow(<Episodes {...props} />);
  return component;
};

describe("Test episodes component", () => {
  describe("Has prop title", () => {
    let wrapper;
    beforeEach(() => {
      const props = {
        title: "All episodes of Ricky and Morty",
      };
      wrapper = setUp(props);
    });

    it("Renders episodes without error", () => {
      const component = findByAttri(wrapper, `episodes-card`);
      expect(component.length).toBe(1);
    });
  });
});
