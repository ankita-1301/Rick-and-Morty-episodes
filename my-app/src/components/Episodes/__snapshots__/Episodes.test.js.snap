// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Episodes component snapshot 1`] = `
<div>
  <div
    className="ant-card ant-card-bordered"
    data-test="episodes-card"
  >
    <div
      className="ant-card-head"
      style={Object {}}
    >
      <div
        className="ant-card-head-wrapper"
      >
        <div
          className="ant-card-head-title"
        >
          All episodes of Ricky and Morty
        </div>
      </div>
    </div>
    <div
      className="ant-card-body"
      style={Object {}}
    />
  </div>
</div>
`;
